File: sample/src/main/java/com/kennyc/sample/MainActivity.java
package com.kennyc.sample;

import android.os.Bundle;
import android.os.Handler;
import android.os.Message;
import android.support.annotation.NonNull;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.ListView;
import android.widget.Toast;

import com.kennyc.view.MultiStateView;


public class MainActivity extends AppCompatActivity implements MultiStateView.StateListener {

    private TestHandler mHandler = new TestHandler();

    private MultiStateView mMultiStateView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        mMultiStateView = (MultiStateView) findViewById(R.id.multiStateView);
        mMultiStateView.setStateListener(this);
        mMultiStateView.getView(MultiStateView.VIEW_STATE_ERROR).findViewById(R.id.retry)
                .setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        mMultiStateView.setViewState(MultiStateView.VIEW_STATE_LOADING);
                        Toast.makeText(getApplicationContext(), "Fetching Data", Toast.LENGTH_SHORT).show();
                        Message msg = mHandler.obtainMessage();
                        msg.obj = mMultiStateView;
                        mHandler.sendMessageDelayed(msg, 3000);
                    }
                });

        ListView list = (ListView) mMultiStateView.findViewById(R.id.list);

        String[] data = new String[100];
        for (int i = 0; i < 100; i++) {
            data[i] = "Row " + i;
        }

        list.setAdapter(new ArrayAdapter<>(getApplicationContext(), android.R.layout.simple_list_item_1, data));
    }

    @Override
    protected void onPause() {
        mHandler.removeCallbacksAndMessages(null);
        super.onPause();
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.menu_main, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.error:
                mMultiStateView.setViewState(MultiStateView.VIEW_STATE_ERROR);
                return true;

            case R.id.empty:
                mMultiStateView.setViewState(MultiStateView.VIEW_STATE_EMPTY);
                return true;

            case R.id.content:
                mMultiStateView.setViewState(MultiStateView.VIEW_STATE_CONTENT);
                return true;

            case R.id.loading:
                mMultiStateView.setViewState(MultiStateView.VIEW_STATE_LOADING);
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onStateChanged(@MultiStateView.ViewState int viewState) {
        Log.v("MSVSample", "onStateChanged; viewState: " + viewState);
    }

    @Override
    public void onStateInflated(@MultiStateView.ViewState int viewState, @NonNull View view) {
        Log.v("MSVSample", "onStateInflated; viewState: " + viewState + ", view: " + view.toString());
    }

    private static class TestHandler extends Handler {

        @Override
        public void handleMessage(Message msg) {
            if (msg.obj instanceof MultiStateView) {
                ((MultiStateView) msg.obj).setViewState(MultiStateView.VIEW_STATE_CONTENT);
            }

            super.handleMessage(msg);
        }
    }
}

File: sample/src/main/java/com/kennyc/sample/MainActivity.kt
package com.kennyc.sample

import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.ListView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import com.kennyc.view.MultiStateView

class MainActivity : AppCompatActivity(), MultiStateView.StateListener {
    private lateinit var multiStateView: MultiStateView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)
        multiStateView = findViewById(R.id.multiStateView)
        multiStateView.listener = this
        multiStateView.getView(MultiStateView.VIEW_STATE_ERROR)?.findViewById<Button>(R.id.retry)
                ?.setOnClickListener {
                    multiStateView.viewState = MultiStateView.VIEW_STATE_LOADING
                    Toast.makeText(applicationContext, "Fetching Data", Toast.LENGTH_SHORT).show()
                    multiStateView.postDelayed({ multiStateView.viewState = MultiStateView.VIEW_STATE_CONTENT }, 3000L)
                }

        val list: ListView = multiStateView.findViewById(R.id.list)

        val data = arrayOfNulls<String>(100)
        for (i in 0..99) {
            data[i] = "Row $i"
        }

        list.adapter = ArrayAdapter<String>(applicationContext, android.R.layout.simple_list_item_1, data)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_main, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            R.id.error -> {
                multiStateView.viewState = MultiStateView.VIEW_STATE_ERROR
                return true
            }

            R.id.empty -> {
                multiStateView.viewState = MultiStateView.VIEW_STATE_EMPTY
                return true
            }

            R.id.content -> {
                multiStateView.viewState = MultiStateView.VIEW_STATE_CONTENT
                return true
            }

            R.id.loading -> {
                multiStateView.viewState = MultiStateView.VIEW_STATE_LOADING
                return true
            }
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onStateChanged(viewState: Int) {
        Log.v("MSVSample", "onStateChanged; viewState: $viewState")
    }
}